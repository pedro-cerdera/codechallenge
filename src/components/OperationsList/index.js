import OperationsList from './OperationsList';

export {OperationsList};
