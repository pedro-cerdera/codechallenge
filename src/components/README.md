
# Components

Here we have some components

- [OperationsList](/src/components/OperationsList/README.md)
- [Account](/src/components/Account/README.md)
