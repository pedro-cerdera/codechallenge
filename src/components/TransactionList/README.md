# TransactionList

Show all user operations in app

## Usage

```js

const transactions = [
    {
     value: 200,
     description: "Transaction 1" 
    },
    {
      value: 300,
      description: "Transaction 2" 
    },
  ];
  
const Example = () => (<TransactionList transactions={transactions} />);
```

### Properties

| Prop                    | Description                      | Required      | Default | Format | 
| ----------------------- | ---------------------------------| ------------- | ------- | ------ |
| **`transactions`**      | Array of all transactions        | true.         | none    | [{value: number, description: string}] |
