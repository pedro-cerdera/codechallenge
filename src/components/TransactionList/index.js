import TransactionList from './TransacitonList';

export {TransactionList};
