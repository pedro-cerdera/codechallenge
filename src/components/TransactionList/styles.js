import {StyleSheet} from 'react-native';
import styleguide from '../../styleguide';

export default StyleSheet.create({
  container: {marginTop: 10},
  separator: {height: 5},
  itemContainer: {flexDirection: 'row', marginHorizontal: 20},
  statusContainer: {marginRight: 20, alignItems: 'center'},
  line: {
    width: StyleSheet.hairlineWidth,
    flex: 1,
    backgroundColor: styleguide.colors.moon400,
    marginTop: 3,
  },
  detailContainer: {flex: 1},
});
