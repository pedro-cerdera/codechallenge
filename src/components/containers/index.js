export * from './SafeAreaContainer';
export * from './CardContainer';
export * from './DefaultScapingContainer';
