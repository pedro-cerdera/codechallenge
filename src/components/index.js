export * from './containers';
export * from './Account';
export * from './Icon';
export * from './OperationsList';
