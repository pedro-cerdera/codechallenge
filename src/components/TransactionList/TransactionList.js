import React from 'react';
import {FlatList, View} from 'react-native';
import PropTypes from 'prop-types';
import {maskMoney} from '../../helpers';
import styleguide from '../../styleguide';
import {CustomText} from '../CustomText';
import {Icon} from '../Icon';
import styles from './styles';

export const TransactionItem = ({value, description}) => (
  <View style={styles.itemContainer}>
    <View style={styles.statusContainer}>
      <Icon
        name={value < 0 ? 'md-money-off' : 'md-attach-money'}
        size={24}
        color={styleguide.colors.moon400}
      />
      <View style={styles.line} />
    </View>
    <View style={styles.detailContainer}>
      <CustomText type={'subtitle'} testID={'operation'}>
        {value < 0 ? 'Saque' : 'Depósito'}
      </CustomText>
      <CustomText testID={'description'}>
        {description || 'Sem descrição'}
      </CustomText>
      <CustomText testID={'value'}>{maskMoney(Math.abs(value))}</CustomText>
    </View>
  </View>
);

const TransactionList = ({transactions}) => (
  <FlatList
    data={transactions}
    renderItem={({item}) => <TransactionItem {...item} />}
    ItemSeparatorComponent={() => <View style={styles.separator} />}
    keyExtractor={(item, index) => `${index}`}
    style={styles.container}
  />
);

TransactionList.propTypes = {
  transactions: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.number.isRequired,
      description: PropTypes.string,
    }),
  ).isRequired,
};

export default TransactionList;
