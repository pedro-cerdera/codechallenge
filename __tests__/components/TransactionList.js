import React from 'react';
import {FlatList} from 'react-native';
import {render} from 'react-native-testing-library';

import {TransactionList} from '../../src/components';
import {TransactionItem} from '../../src/components/TransactionList/TransactionList';

jest.mock('../../src/components/Icon/Icon', () => 'Icon');

const mockDataTransaction = [
  {
    value: -100,
    description: 'Transação 1',
  },
  {
    value: 2100,
    description: 'Transação 2',
  },
];

describe('Testing Transactions', () => {
  test('Render all transactions Itens correctly', () => {
    const componentTree = render(
      <TransactionList transactions={mockDataTransaction} />,
    );

    expect(componentTree.UNSAFE_getAllByType(FlatList).length).toBe(1);
    expect(componentTree.UNSAFE_getAllByType(TransactionItem).length).toBe(
      mockDataTransaction.length,
    );
  });

  test('Test all transactions is correctly', () => {
    const {getAllByTestId} = render(
      <TransactionList transactions={mockDataTransaction} />,
    );

    expect(getAllByTestId('operation')[0].children[0]).toBe('Saque');
    expect(getAllByTestId('operation')[1].children[0]).toBe('Depósito');

    expect(getAllByTestId('value')[0].children[0]).toBe('R$ 100,00');
    expect(getAllByTestId('value')[1].children[0]).toBe('R$ 2.100,00');

    expect(getAllByTestId('description')[0].children[0]).toBe('Transação 1');
    expect(getAllByTestId('description')[1].children[0]).toBe('Transação 2');
  });

  test('renders correctly', () => {
    const tree = render(
      <TransactionList transactions={mockDataTransaction} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
