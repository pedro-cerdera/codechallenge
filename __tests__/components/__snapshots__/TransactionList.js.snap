// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Transactions renders correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  data={
    Array [
      Object {
        "description": "Transação 1",
        "value": -100,
      },
      Object {
        "description": "Transação 2",
        "value": 2100,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "marginTop": 10,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginHorizontal": 20,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginRight": 20,
            }
          }
        >
          <Icon
            color="#8296a4"
            name="md-money-off"
            size={24}
          />
          <View
            style={
              Object {
                "backgroundColor": "#8296a4",
                "flex": 1,
                "marginTop": 3,
                "width": 0.5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#495c69",
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="operation"
          >
            Saque
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#597183",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="description"
          >
            Transação 1
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#597183",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="value"
          >
            R$ 100,00
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "height": 5,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginHorizontal": 20,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "marginRight": 20,
            }
          }
        >
          <Icon
            color="#8296a4"
            name="md-attach-money"
            size={24}
          />
          <View
            style={
              Object {
                "backgroundColor": "#8296a4",
                "flex": 1,
                "marginTop": 3,
                "width": 0.5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#495c69",
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="operation"
          >
            Depósito
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#597183",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="description"
          >
            Transação 2
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#597183",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 18,
                },
                undefined,
              ]
            }
            testID="value"
          >
            R$ 2.100,00
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
